### PR TITLE
implement Haptic Socket and plug channels

### PR DESCRIPTION
**Pull request by TayouVR**
these can be animated. Animating a socket channel will alter its light range. 

~~auto generating a menu for the channels would also be cool, but I didn't really know where to put it, so I opted for this instead... maybe making it a auto generated menu item is better though~~ 
did the auto generated menu too, though I think this isn't quite working rn, not sure why.. 
I wanted to make it also possible to animate the ring/hole property, but it got a bit messy in the code, so i reverted and let it at this. I hope i can get that also animated late down the line somehow

supersedes https://github.com/VRCFury/VRCFury/pull/43

This currently only does the original channel 1 from DPS as the alpha based channels are a bit weird to implement due to their quantitiy....

```
The changes made in this contribution are free and unencumbered software
released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <https://unlicense.org>
```

### CODE REVIEW

**Comment by SenkyDragon** [comment]
We've mostly abandoned the alpha-based channels because they do not work properly on the light in slot 4 due to some unity bug (the color value of the slot 4 light is basically always corrupt).

I'm a bit reluctant to add channel 1 support directly as is, because should we slice it up into multiple channels? The DPS ranges were quite large, we could probably fit 5 or so channels in the same ranges that the old lights used.

**Comment by TayouVR** [comment]
the question is; 
is that useful? Are those channels even gonna be usable at all?
Like; two channels make sense as you can see two sockets at a time. You can't see 3, 4 or 5 at a time though, so supporting that many channels wouldn't really help, would it?
the second channel is only to force a plug to go into a specific socket, like when you have two sockets quite close together (e.g. between the legs) and want to use a specific one for each plug. Then two of the three players can switch a pair of socket and plug to ch1 while the other pair stays on channel 0 (default)

File: com.vrcfury.vrcfury/SPS/sps_search.cginc [hunk 1]
```diff
@@ -29,6 +29,13 @@ void sps_parse_light(float range, out int type, int myChannel) {
 		if (legacyRange == 5) type = 3;
 	// }
 
+	// Handle legacy DPS channel 1
+	if (myChannel == -1)	{
```
Comment (SenkyDragon) [inline com.vrcfury.vrcfury/SPS/sps_search.cginc:33]: I just noticed, this makes it work with BOTH channel 0 and channel 1, which is probably not right

Comment (TayouVR) [reply com.vrcfury.vrcfury/SPS/sps_search.cginc]: oops, should be fixed now

**Comment by TayouVR** [comment]
rebased and cleaned up a bit

File: com.vrcfury.vrcfury/SPS/sps_search.cginc [hunk 2]
```diff
@@ -24,9 +24,15 @@ void sps_parse_light(float range, half4 color, out int type, int myChannel) {
...
+			if (legacyRange == 4) type = 2;
+			if (legacyRange == 6) type = 3;
+		} else { // Default Channel
+			if (legacyRange == 1) type = 1;
+			if (legacyRange == 2) type = 2;
+			if (legacyRange == 5) type = 3;
+		}
 	// }
 
 	if (channel != myChannel) {
```
Comment (TheLastRar) [inline com.vrcfury.vrcfury/SPS/sps_search.cginc:38]: Just having a quick look, but wouldn't this condition always fail with DPS channel 1?
`myChannel`  would be `-1`, but `channel` is initialised to 0 on line 15, and never updated

Unless I missed something and this works regardless?

**Comment by TayouVR** [comment]
oh, I didn't realize rebasing makes the "Compare" thing basically useless. maybe for PRs I should just do merge commits and keep fixes in separate commits to make it easier to see whats going on...

**Comment by SenkyDragon** [comment]
I'm going to give you credit in CONTRIBUTING.md, but I don't think I'm going to merge this. All efforts will be focused on lightless SPS, which will likely support its own channel system that does not depend on lights.

Since we do not really need to depend on cross-compatibility as DPS channels were so rarely used, I feel like this is an acceptable tradeoff, since the dedicated few that were using it could move to lightless SPS channels.

**Pull request closed without merging**